rotkehlchen/externalapis/covalent.py

```diff
@@ -0,0 +1,248 @@
+log = RotkehlchenLogsAdapter(logger)
+
+
+def read_hash(data: Dict[str, Any], key: str) -> bytes:
+    try:
+        result = hexstring_to_bytes(data[key])
+    except ValueError as e:
+        raise DeserializationError(
+            f'Failed to read {key} as a hash during Covalent transaction query',
+        ) from e
+    return result
+
+
+def read_integer(data: Dict[str, Any], key: str) -> int:
+    try:
+        result = convert_to_int(data[key])
+    except ConversionError as e:
+        raise DeserializationError(
+            f'Failed to read {key} as an integer during Covalent transaction query',
+        ) from e
+    return result
+
+
+def convert_transaction_from_covalent(
+        data: Dict[str, Any],
+) -> CovalentTransaction:
+    """Reads dict data of a transaction from Covalent and deserializes it
+
+    Can raise DeserializationError if something is wrong
+    """
+    try:
+        # internal tx list contains no gasprice
+        timestamp = datetime.timestamp(
+            datetime.strptime(data['block_signed_at'], '%Y-%m-%dT%H:%M:%SZ'),
+        )
+        timestamp = int(timestamp)
+        return CovalentTransaction(
+            timestamp=Timestamp(timestamp),
+            block_number=data['block_height'],
+            tx_hash=data['tx_hash'],
+            from_address=data['from_address'],
+            to_address=data['to_address'],
+            value=read_integer(data, 'value'),
+            gas=read_integer(data, 'gas_offered'),
+            gas_price=read_integer(data, 'gas_price'),
+            gas_used=read_integer(data, 'gas_spent'),
+            input_data='0x',
+            nonce=0,
+        )
+    except KeyError as e:
+        raise DeserializationError(
+            f'Covalent avalanche transaction missing expected key {str(e)}',
+        ) from e
+
+
+class Covalent():
+    def __init__(self,
+                 msg_aggregator: MessagesAggregator,
+                 chain_id: str) -> None:
+        self.session = requests.session()
+        self.session.headers.update({'User-Agent': 'rotkehlchen'})
+        self.msg_aggregator = msg_aggregator
+        self.chain_id = chain_id
+
+    def _query(
+            self,
+            module: str,
+            action: str,
+            address: str = None,
+            options: Optional[Dict[str, Any]] = None,
+            timeout: Optional[Tuple[int, int]] = None,
+    ) -> Any:
+        """Queries Covalent
+
+        May raise:
+        - RemoteError if there are any problems with reaching Covalent or if
+        an unexpected response is returned
+        """
+        query_str = f'https://api.covalenthq.com/v1/{self.chain_id}/{action}'
+        if address:
+            query_str += f'/{address}'
+        query_str += f'/{module}/'
+
+        query_str += f'?key={KEY}'
+
+        if options:
+            for name, value in options.items():
+                query_str += f'&{name}={value}'
+
+        backoff = 1
+        backoff_limit = 33
```

Comment:
They just raise a 409 (Too Many Requests), but the API is pretty sturdy this only tend to happen in multiple race conditions, we gonna change that to just do a retry in case we get a timeout from them (which is more common).